Jsx (#128)
reactjs/es.reactjs.org: content/docs/jsx-in-depth.md

  - "docs/jsx-in-depth-ko-KR.html"
---

Fundamentalmente, JSX solo proporciona azúcar sintáctica para la función `React.createElement(component, props, ...children)`. El código JSX:

```js
<MyButton color="blue" shadowSize={2}>

>   Haz click en mi

Reviewer: ```suggestion
  Haz click en mí
```

</MyButton>
```

compila en:

```js
React.createElement(
  MyButton,
  {color: 'blue', shadowSize: 2},

>   'Haz click en mi'

Reviewer: ```suggestion
  'Haz click en mí'
```

}
```

Puedes obtener más información sobre [representación condicional](/docs/conditional-rendering.html) y [loops](/docs/lists-and-keys.html) en las secciones correspondientes.

> ### Literales String {#string-literals}

Reviewer: ```suggestion
### Literales *string* {#string-literals}
```

Puede pasar una string literal como prop. Estas dos expresiones JSX son equivalentes:

```js
<MyComponent message="hello world" />

<MyComponent message={'hello world'} />
```

Cuando se pasa un literal string, su valor no se escapa de HTML. Así que estas dos expresiones JSX son equivalentes:

```js
<MyComponent message="&lt;3" />

<MyComponent message={'<3'} />
```

Este comportamiento no suele ser relevante. Sólo se menciona aquí para completar.

> ### Props Predeterminado a "Verdadero" {#props-default-to-true}

Reviewer: El valor `verdadero` no existe en Javascript, por lo que en este caso no se debe traducir. Dejé props en masculino como lo has usado hasta ahora, porque no se ha decidio aún si deban ir en femenino al ser una forma abreviada de propiedades.
```suggestion
### Los props son por defecto "true" {#props-default-to-true}
```

Los atributos de propagación pueden ser útiles, pero también facilitan la transferencia de propiedades innecesarias a los componentes que no les interesan o la transferencia de atributos HTML no válidos al DOM. Recomendamos utilizar esta sintaxis con moderación. 

## Niños en JSX {#children-in-jsx}

En las expresiones JSX que contienen una etiqueta de apertura y una etiqueta de cierre, el contenido entre esas etiquetas se pasa como un elemento especial: `props.children`. Hay varias maneras diferentes de pasar a los niños:

### Cadenas de Literales {#string-literals-1}

Puede poner una cadena entre las etiquetas de apertura y cierre y `props.children` será solo esa cadena. Esto es útil para muchos de los elementos HTML integrados. Por ejemplo:

```js
<MyComponent>Hello world!</MyComponent>
```

Esto es JSX válido, y `props.children` en` MyComponent` simplemente será la cadena `Hello world!`. HTML no se ha escapado, por lo que generalmente puedes escribir JSX de la misma manera que escribiría HTML de esta manera:

```html
<div>This is valid HTML &amp; JSX at the same time.</div>
```

> JSX elimina los espacios en blanco al principio y al final de una línea. También elimina las líneas en blanco. Se eliminan las nuevas líneas adyacentes a las etiquetas; Las nuevas líneas que se producen en medio de cadenas literales se condensan en un solo espacio. Así que todo esto se traduce en lo mismo:

Reviewer: ```suggestion
JSX elimina los espacios en blanco al principio y al final de una línea. También elimina las líneas en blanco. Se eliminan las nuevas líneas adyacentes a las etiquetas, las nuevas líneas que se producen en medio de literales de *string* se condensan en un solo espacio. Así que todo esto se traduce en lo mismo:
```

</div>
```

> ### JSX Niños {#jsx-children}

Reviewer: ```suggestion
### Los hijos JSX {#jsx-children}
```

> Puede proporcionar más elementos JSX como los niños. Esto es útil para mostrar componentes anidados:

Reviewer: ```suggestion
Puedes proporcionar más elementos JSX como hijos. Esto es útil para mostrar componentes anidados:
```

</MyContainer>
```

> Puedes mezclar diferentes tipos de niños, así que puedes usar literales de string junto con niños JSX. Esta es otra forma en la que JSX es como HTML, de modo que es tanto JSX válido como HTML válido:

Reviewer: ```suggestion
Puedes mezclar diferentes tipos de hijos, así que puedes usar literales de *string* junto con hijos JSX. Esta es otra forma en la que JSX es como HTML, de modo que es tanto JSX válido como HTML válido:
```

```js
render() {
  // ¡No es necesario envolver los elementos de la lista en un elemento adicional!
  return [
    // No olvides las llaves :)
    <li key="A">First item</li>,
    <li key="B">Second item</li>,
    <li key="C">Third item</li>,
  ];
}
```

> ### Expresiones de JavaScript como niños {#javascript-expressions-as-children}

Reviewer: ```suggestion
### Expresiones de JavaScript como hijos {#javascript-expressions-as-children}
```

> Puede pasar cualquier expresión de JavaScript como hijos, encerrándola dentro de `{}`. Por ejemplo, estas expresiones son equivalentes:

Reviewer: ```suggestion
Puedes pasar cualquier expresión de JavaScript como hijos, encerrándola dentro de `{}`. Por ejemplo, estas expresiones son equivalentes:
```

}
```

> Las expresiones de JavaScript se pueden mezclar con otros tipos de niños. Esto suele ser útil en lugar de las plantillas de cadena:

Reviewer: ```suggestion
Las expresiones de JavaScript se pueden mezclar con otros tipos de hijos. Esto suele ser útil en lugar de *string templates*:
```

```js{2}
function Hello(props) {
  return <div>Hello {props.addressee}!</div>;
}
```

> ### Funciones como niños {#functions-as-children}

Reviewer: ```suggestion
### Funciones como hijos {#functions-as-children}
```

> Normalmente, las expresiones de JavaScript insertadas en JSX se evaluarán como una cadena, un elemento React o una lista de esas cosas. Sin embargo, `props.children` funciona igual que cualquier otro prop, ya que puede pasar cualquier tipo de datos, no solo los tipos que React sabe cómo procesar. Por ejemplo, si tiene un componente personalizado, puedes hacer que tome un callback como `props.children`:

Reviewer: ```suggestion
Normalmente, las expresiones de JavaScript insertadas en JSX se evaluarán como una cadena, un elemento React o una lista de esas cosas. Sin embargo, `props.children` funciona igual que cualquier otro prop, ya que puede pasar cualquier tipo de datos, no solo los tipos que React sabe cómo procesar. Por ejemplo, si tienes un componente personalizado, puedes hacer que tome un *callback* como `props.children`:
```

```js{4,13}

> // Llama al callback de los elementos secundarios para producir un componente repetido

Reviewer: ```suggestion
// Llama al *callback* `numTimes` veces para producir un componente repetido
```

}
```

Los niños pasados a un componente personalizado pueden ser cualquier cosa, siempre que ese componente los transforme en algo que React pueda entender antes de renderizar. Este uso no es común, pero funciona si desea ampliar lo que JSX es capaz de hacer.

> ### Vooleanos, Nulos e Indefinidos se ignoran {#booleans-null-and-undefined-are-ignored}

Reviewer: ```suggestion
### Los booleanos, `null` y `undefined` se ignoran {#booleans-null-and-undefined-are-ignored}
```

</div>
```

> Una advertencia es que algunos [valores "falsos"](https://developer.mozilla.org/en-US/docs/Glossary/Falsy), como el número `0`, todavía son renderizados por React. Por ejemplo, este código no se comportará como se espera porque se imprimirá `0` cuando `props.messages` sea una arreglo vacío:

Reviewer: Los enlaces a MDN deben ser los correspondientes a la edición en español.
```suggestion
Una advertencia es que algunos [valores que se comportan como falsos](https://developer.mozilla.org/es/docs/Glossary/Falsy), como el número `0`, todavía son renderizados por React. Por ejemplo, este código no se comportará como se espera porque se imprimirá `0` cuando `props.messages` sea una arreglo vacío:
```

</div>
```

> Para solucionar esto, asegúrese de que la expresión antes de `&&` sea siempre booleana:

Reviewer: ```suggestion
Para solucionar esto, asegúrate de que la expresión antes de `&&` sea siempre booleana:
```

```

> Conversely, if you want a value like `false`, `true`, `null`, or `undefined` to appear in the output, you have to [convert it to a string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#String_conversion) first:

Reviewer: ```suggestion
```

> Por el contrario, si desea que aparezca un valor como `false`, `true`, `null` o `undefined` en la salida, debe [convertirlo en una cadena](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#String_conversion) primero:

Reviewer: ```suggestion
Por el contrario, si desea que aparezca un valor como `false`, `true`, `null` o `undefined` en la salida, debes primero [convertirlo en una cadena](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#String_conversion):
```
